File: src/Json.fs
module Aesop.JsonParser

open Fable.Core

type JsonType =
    | String of string
    | Number of float
    | Object of array<string * JsonType>
    | Array of array<JsonType>
    | Boolean of bool
    | Null

    
// TODO implement this in F# rather than the FFI
[<Import("ofString", "./Json.js")>]
let private _ofString:
    (string -> JsonType) ->
    (float -> JsonType) ->
    (array<string * JsonType> -> JsonType) ->
    (array<JsonType> -> JsonType) ->
    (bool -> JsonType) ->
    JsonType ->
    (JsonType -> Result<JsonType, System.Exception>) ->
    (System.Exception -> Result<JsonType, System.Exception>) ->
    (string -> Result<JsonType, System.Exception>) = jsNative

let ofString: string -> Result<JsonType, System.Exception> = _ofString String Number Object Array Boolean Null Ok Error

File: src/json.js
const getType = (a) =>
  ({}).toString.call(a);

const mapObject = (a, f) =>
  Object.keys(a).map((key) => f(key, a[key]));

export const ofString = (String, Number, Object, Array, Boolean, Null, Some, None) => {
  const makeJson = (a) => {
    switch (getType(a)) {
    case "[object Object]":
      return Some(Object(mapObject(a, (key, value) => [key, makeJson(value)])));
    case "[object Array]":
      return Some(Array(a.map(makeJson)));
    case "[object Number]":
      return Some(Number(a));
    case "[object String]":
      return Some(String(a));
    case "[object Boolean]":
      return Some(Boolean(a));
    case "[object Null]":
      return Some(Null);
    default:
      return None;
    }
  };

  return (input) => {
    try {
      return makeJson(JSON.parse(input));

    // TODO maybe return the error message ?
    } catch (e) {
      return None;
    }
  };
};